Add `colors` support for `tf.image.draw_bounding_boxes`

yongtang opened the pull request:
This fix tries to address the issue raised in #15692 where
it was not possible to specify the colors for boxes in
`tf.image.draw_bounding_boxes`. Instead, a predefined
fixed color table was used to cycle through colors.

This fix adds `colors` Input to `DrawBoundingBoxexV2`
so that it is possible to specify the color. In case
no color is specified, the default color table will
be used.

Since there is an API change, the op is labeled as V2.

This fix fixes #15692.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `colors` support for `tf.image.draw_bounding_boxes`

This fix tries to address the issue raised in 15692 where
it was not possible to specify the colors for boxes in
`tf.image.draw_bounding_boxes`. Instead, a predefined
fixed color table was used to cycle through colors.

This fix adds `colors` Input to `DrawBoundingBoxexV2`
so that it is possible to specify the color. In case
no color is specified, the default color table will
be used.

Since there is an API change, the op is labeled as V2.

This fix fixes 15692.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Register `DrawBoundingBoxV2` op in image_ops.cc

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Sanitize with clang-format -i --style=Google

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for colors support for `gen_image_ops.draw_bounding_box_v2`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update api defs

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Replace std::array with std::vector to address Mac build issues

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Hide draw_bounding_boxes_v2 first

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflowbutler commented:
Nagging Reviewer @jaingaurav: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

jaingaurav reviewed (commented):
Thanks for the PR. Any reason why we should make colors an input argument vs an attribute? Is it useful to send the colors as a list of tensors? Advantage of using an attribute is we won't need to add a new V2 op. Using an attribute might also help avoid the copying of the color vector in draw_bounding_box_op.cc.

yongtang commented:
@asimshankar @jaingaurav I take a look at the `colors` field, It is actually a list of RGBA colors so it is 2-D. Not sure it would be suitable for `Attr`. Should this be wrapped into 1-D (4 * colors) instead?

asimshankar commented:
Attribute values can be tensors as well (though the only real example of that is the [`Const` op](https://github.com/tensorflow/tensorflow/blob/c842394896de32c4160fc0f5f94dabf06890476d/tensorflow/core/ops/array_ops.cc#L697)), so you could still use an attribute.

Now, whether this should be an attribute or an input pretty much depends on whether the list of colors is something that is computed in the graph or is static. The former is certainly more general.

Shouldn't this PR also set things up so that `tf.image.draw_bounding_boxes` has an `colors` argument?

asimshankar commented:
(From API review): Using an input seems appropriate for this. So, please go ahead and update `tf.image.draw_bounding_boxes` so that it uses the V2 op if there say a non-`None` `colors` argument or something like that. Thanks @yongtang !

yongtang committed:
Update to expose image.draw_bounding_boxes in v1 and v2 differently

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@asimshankar Thanks for the review. The PR has been updated with v1 pointing to old behavior and v2 allows colors input. Let me know if this is the expected implementation.

asimshankar commented on tensorflow/python/kernel_tests/draw_bounding_box_op_test.py:
```diff
@@ -85,9 +90,12 @@ def _testDrawBoundingBoxColorCycling(self, img):
       image = image_ops_impl.convert_image_dtype(image, dtypes.float32)
       image = array_ops.expand_dims(image, 0)
-      image = image_ops.draw_bounding_boxes(image, bboxes)
+      if colors is None:
```
This shouldn't be needed. We should be testing the `draw_bounding_boxes` function exposed in the API instead of testing the generated wrappers. So, can't this be:

```python
image = image_ops.draw_bounding_boxes(image, bboxes, colors=colors_arg)
```
?

Where `colors_arg` points to the `colors` value provided to `_testBoundingBoxColorCycling` (i.e., before it was set to `default_color_table` to generate `test_drawn_image`?

yongtang replied:
Updated.

asimshankar commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3057,3 +3057,70 @@ def crop_and_resize_v2(
+  return gen_image_ops.draw_bounding_boxes(images, boxes, name)
+
+@tf_export('image.draw_bounding_boxes', v1=[])
+def draw_bounding_boxes_v2(
```
Might as well keep the colors functionality in both V1 and V2, no?
Something like:

```python
@tf_export('image.draw_bounding_boxes', v1=[])
def draw_bounding_boxes_v2(images, boxes, colors=None, name=None):
  if colors is None and not compat.forward_compatible(2019, 05, 01):
    return gen_image_ops.draw_bounding_boxes(images, boxes, name)
  return gen_image_ops.draw_bounding_boxes_v2(images, boxes, colors, name)

@tf_export(v1=['image.draw_bounding_boxes'])
def draw_bounding_boxes(images, boxes, name=None, colors=None):
  return draw_bounding_boxes_v2(images, boxes, colors, name)
```

We don't need separate wrappers for each of the generated functions.
Thoughts?

Where `compat.forward_compatible` is from: https://github.com/tensorflow/tensorflow/blob/2f2503d5d0c27104848fcee418ca9f97607cc39c/tensorflow/python/compat/compat.py#L34

yongtang replied:
Updated.

asimshankar reviewed (changes requested):
Apologies for the late reply, some suggestions.

yongtang committed:
Keep the colors functionality for both v1 and v2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Use DefaultColorTable utility function

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix pylint error of `Unused gen_image_ops imported `

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/kernels/draw_bounding_box_op.cc:
```diff
@@ -64,19 +64,49 @@ class DrawBoundingBoxesOp : public OpKernel {
     };
 
+    std::vector<std::vector<float>> color_table;
+    for (int64 i = 0; i < default_color_table_length; i++) {
```
Thanks. Done.

yongtang commented on tensorflow/core/kernels/draw_bounding_box_op.cc:
```diff
@@ -64,19 +64,49 @@ class DrawBoundingBoxesOp : public OpKernel {
+      OP_REQUIRES(context, colors_tensor.shape().dims() == 2,
+                  errors::InvalidArgument("colors must be a 2-D matrix",
+                                          colors_tensor.shape().DebugString()));
+      OP_REQUIRES(context, colors_tensor.shape().dim_size(1) == 4,
```
Thanks. The check has been updated. Now it requires the number of channels >= needed channels.

yongtang commented:
Thanks @asimshankar for the review. The PR has been updated. Please take a look.

hgadig commented:
@yongtang  Could you please look into the merging conflicts it has. Thanks !

yongtang commented:
Thanks @hgadig for the help. The PR has been rebased with merge conflict resolved. Please take a look and let me know if there are any other issues.

hgadig commented:
> Thanks @hgadig for the help. The PR has been rebased with merge conflict resolved. Please take a look and let me know if there are any other issues.

Thanks for fixing. Waiting for the approval by reviewer.

hgadig commented:
@yongtang  Not sure why there are branch conflicts again after doing rebase. 
@alextp  Can we create a new PR for the same(by closing this) ? Or any other good suggestion to bypass this branch conflicts ? Please note that @yongtang  has already rebased.

yongtang commented:
@hgadig There were some changes in image_ops_impl.py (`combined_non_max_suppression`) which caused the merge conflict. Guess the `git diff` is not smart enough to find out they are unrelated to this PR. I have rebased and updated the PR to resolve the merge conflict. Please take a look.

hgadig commented:
> @hgadig There were some changes in image_ops_impl.py (`combined_non_max_suppression`) which caused the merge conflict. Guess the `git diff` is not smart enough to find out they are unrelated to this PR. I have rebased and updated the PR to resolve the merge conflict. Please take a look.

Thank you !

@alextp  Could you please approve it.

yongtang committed:
Fix api compat issue

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Ping @hgadig to take a look. I rebased the PR as otherwise there is a merge conflict. Since the PR has been approved, wondering if there is anything else from that needs to be done? Would like to see this PR land soon.

hgadig commented:
> Ping @hgadig to take a look. I rebased the PR as otherwise there is a merge conflict. Since the PR has been approved, wondering if there is anything else from that needs to be done? Would like to see this PR land soon.

I do not see any approved(green tick mark) sign yet. Unless it is shown, I cannot proceed with the next steps.

yongtang commented:
Thanks @hgadig for the help. I think @alextp already approved at one point. However, since there were merge conflicts, I rebased and pushed the PR several times. With the rebase, the approval comments were overridden.

@alextp Could you take a look to see if everything is ok, and re-approve if it is fine?

tensorflowbutler commented:
Nagging Reviewer @jaingaurav, @asimshankar: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

yongtang commented:
/cc @alextp to take a look. I rebased and updated PR of the last commit, as there were some changes in compat test recently.

I think now all tests pass. The sanity test failed but it is unrelated and is being addressed in PR #27226.
